File: Evaluator.cpp
/**
 * @Algorithm
 * Shunting Yard algorithm created by Edsger W. Dijkstra.
 * @License
 * This calculator uses material from the Wikipedia article
 * Shunting-yard algorithm (https://en.wikipedia.org/wiki/Shunting-yard_algorithm),
 * which is released under the
 * Creative Commons Attribution-Share-Alike License 3.0 (https://creativecommons.org/licenses/by-sa/3.0/)
 */

#include <iostream> // cout && getline
#include <string>   // string
#include <vector>   // vector
#include <deque>    // deque
#include <cmath>    // pow

enum class tokenType {
    nil,
    lpa,
    rpa,
    add,
    sub,
    mul,
    div,
    mod,
    exp,
    i32,
    f32
};

class token {
public:
    [[nodiscard]] std::string toString() const {
        return "(" + strData + ", " + std::to_string(intData)
                + ", " + std::to_string(fltData) + ", [" + std::to_string(precedence)
                + " : " + tokenTypeStrings[(int)type] + "])";
    }

    std::string strData {};
    long intData {};
    float fltData {};
    size_t precedence {};
    tokenType type {tokenType::nil};
    bool rAssociative {false};
    bool unary {false};

private:
    const std::vector<const char *> tokenTypeStrings {
            "nil",
            "lpa",
            "rpa",
            "add",
            "sub",
            "mul",
            "div",
            "mod",
            "exp",
            "i32",
            "f32"
    };
};

class lexana {
public:
    /* For debugging lexer output. */
    [[maybe_unused]] [[nodiscard]] std::string toString() const {
        std::string s {"[\n"};
        for (const token &t: formatTokens) {
            s += '\t' + t.toString() + '\n';
        }
        s += ']';
        return s;
    }

    [[nodiscard]] std::vector<token> &getTokens() {
        return formatTokens;
    }

    void lex(const std::string &_data) {
        data = _data;

        for (size_t i {0}; i < data.size(); ++i) {
        _again:
            char c {data[i]};
            token t {};

            switch (c) {
                case ' ':
                case '\t':
                case '\n':
                case '\r':
                    ++i;
                    goto _again;

                case '(':
                    t.type = tokenType::lpa;
                    t.strData = '(';
                    t.precedence = 9;
                    t.rAssociative = false;
                    break;

                case ')':
                    t.type = tokenType::rpa;
                    t.strData = ')';
                    t.precedence = 0;
                    t.rAssociative = false;
                    break;

                case '+':
                    t.type = tokenType::add;
                    t.strData = '+';
                    t.precedence = 2;
                    t.rAssociative = false;
                    break;

                case '-':
                    if (formatTokens.empty() || (formatTokens.back().type == tokenType::add
                                                 || formatTokens.back().type == tokenType::sub
                                                 || formatTokens.back().type == tokenType::div
                                                 || formatTokens.back().type == tokenType::mul
                                                 || formatTokens.back().type == tokenType::mod
                                                 || formatTokens.back().type == tokenType::exp)
                        || formatTokens.back().type == tokenType::lpa) {
                        t.unary = true;
                        t.precedence = 5;
                        t.rAssociative = false;
                        t.strData = 'm';
                        t.precedence = 5;
                        t.type = tokenType::sub;
                    }
                    else {
                        t.type = tokenType::sub;
                        t.strData = '-';
                        t.precedence = 2;
                        t.rAssociative = false;
                    }
                    break;

                case '/':
                    t.type = tokenType::div;
                    t.strData = '/';
                    t.precedence = 3;
                    t.rAssociative = false;
                    break;

                case '*':
                case 'x':
                    t.type = tokenType::mul;
                    t.strData = '*';
                    t.precedence = 3;
                    t.rAssociative = false;
                    break;

                case '%':
                    t.type = tokenType::mod;
                    t.strData = '%';
                    t.precedence = 6;
                    t.rAssociative = false;
                    break;

                case '^':
                    t.type = tokenType::exp;
                    t.strData = '^';
                    t.precedence = 4;
                    t.rAssociative = true;
                    break;

                case '0':
                case '9':
                case '8':
                case '7':
                case '6':
                case '5':
                case '4':
                case '3':
                case '2':
                case '1': {
                    std::string floatStr {};
                    t.type = tokenType::i32;

                    while (std::isdigit(c) || c == '_' || c == '.') {
                        if (c == '_') {
                            c = data[++i];
                            continue;
                        }
                        else if (c == '.') {
                            if (t.type == tokenType::f32) {
                                std::cerr << "Redefinition of float: " << t.toString() << '\n';
                                exit(1);
                            }
                            else {
                                t.type = tokenType::f32;
                                floatStr += '.';
                            }
                        }
                        else {
                            floatStr += c;
                        }

                        c = data[++i];
                    }

                    if (t.type == tokenType::f32) {
                        t.fltData = std::stof(floatStr);
                    }
                    else {
                        t.intData = std::stoi(floatStr);
                    }

                    --i; // Let for loop skip last char.

                    break;
                }

                case '.':
                    if (isdigit(data[i+1])) {
                        t.type = tokenType::f32;
                        c = data[++i];
                        std::string floatStr {'.'};

                        while (isdigit(c) || c == '_' || c == '.') {
                            if (c == '_') {
                                c = data[++i];
                                continue;
                            }
                            else if (c == '.') {
                                if (t.type == tokenType::f32) {
                                    std::cerr << "Redefinition of float: " << t.toString() << '\n';
                                    exit(1);
                                }
                                else {
                                    t.type = tokenType::f32;
                                    floatStr += '.';
                                }
                            }
                            else {
                                floatStr += c;
                            }

                            c = data[++i];
                        }

                        if (t.type == tokenType::f32) {
                            t.fltData = std::stof(floatStr);
                        }
                        else {
                            t.intData = std::stoi(floatStr);
                        }

                        --i; // Let for loop skip last char.
                    }
                    else {
                        std::cerr << "Unexpected: '.'\n";
                        exit(1);
                    }
                    break;

                default:
                    std::cerr << "Unexpected: " << c << '\n';
                    exit(1);
            }

            formatTokens.push_back(t);
        }
    }

private:
    std::string data {};
    std::vector<token> formatTokens {};
};

std::deque<token> shuntingYard(std::vector<token> &tokens) {
    std::deque<token> queue;
    std::vector<token> stack;

    for (const auto &t: tokens) {
        switch(t.type) {
            case tokenType::i32:
            case tokenType::f32:
                queue.push_back(t);
                break;

            case tokenType::add:
            case tokenType::sub:
            case tokenType::mul:
            case tokenType::div:
            case tokenType::mod:
            case tokenType::exp: {
                const auto o1 {t};

                while(!stack.empty()) {
                    const auto o2 = stack.back();

                    if((! o1.rAssociative && o1.precedence <= o2.precedence)
                       || (o1.rAssociative && o1.precedence <  o2.precedence)) {
                        stack.pop_back();
                        queue.push_back(o2);
                        continue;
                    }

                    break;
                }

                stack.push_back(o1);
                break;
            }

            case tokenType::lpa:
                stack.push_back(t);
                break;

            case tokenType::rpa: {
                bool match {false};

                while (!stack.empty() && stack.back().type != tokenType::lpa) {
                    queue.push_back(stack.back());
                    stack.pop_back();
                    match = true;
                }

                if (!match && stack.empty()) {
                    std::cerr << "Right parenthesis error: " << t.toString() << '\n';
                    return {};
                }

                if (!stack.empty()) stack.pop_back();
                break;
            }

            default:
                std::cerr << "Error: " << t.toString();
                return {};
        }
    }

    while(!stack.empty()) {
        if(stack.back().type == tokenType::lpa) {
            std::cerr << "Mismatched parentheses error\n";
            return {};
        }

        queue.push_back(std::move(stack.back()));
        stack.pop_back();
    }

    return queue;
}

float compute(std::deque<token> &formatted) {
    std::vector<float> stack {};

    while (!formatted.empty()) {
        const token &t {formatted.front()};
        formatted.pop_front();
        switch (t.type) {
            case tokenType::nil: break;

            case tokenType::i32:
                stack.push_back((float)t.intData);
                break;

            case tokenType::f32:
                stack.push_back(t.fltData);
                break;

            case tokenType::add:
            case tokenType::sub:
            case tokenType::mul:
            case tokenType::div:
            case tokenType::mod:
            case tokenType::exp:
                if (t.unary) {
                    float &rhs {stack.back()};
                    rhs *= -1;
                }
                else {
                    const float rhs {stack.back()};
                    stack.pop_back();
                    const float lhs {stack.back()};
                    stack.pop_back();

                    switch (t.type) {
                        case tokenType::exp:
                            stack.push_back(std::pow(lhs, rhs));
                            break;

                        case tokenType::mul:
                            stack.push_back(lhs * rhs);
                            break;

                        case tokenType::div:
                            stack.push_back(lhs / rhs);
                            break;

                        case tokenType::add:
                            stack.push_back(lhs + rhs);
                            break;

                        case tokenType::sub:
                            stack.push_back(lhs - rhs);
                            break;

                        default: break;
                    }
                }
                break;

            default: break;
        }
    }

    return stack.back();
}

int main() {
    std::string exprStr {};
    lexana *lexer {new lexana {}};

    while (true) {
        std::cout << "Enter an mathematical expression ('exit' to stop): ";
        std::getline(std::cin, exprStr);
        std::cout << '\n';

        if (exprStr == "exit") {
            break;
        }

        lexer->lex(exprStr);

        std::deque<token> formatted {shuntingYard(lexer->getTokens())};

        const float expr {compute(formatted)};

        std::cout << "That evaluates out to:\n" << expr << "\n\n";

        lexer->getTokens().clear();
    }

    delete lexer;

    return 0;
}

File: Evaluator.go
/**
 * @Algorithm
 * Shunting Yard algorithm created by Edsger W. Dijkstra.
 * @License
 * This calculator uses material from the Wikipedia article
 * Shunting-yard algorithm (https://en.wikipedia.org/wiki/Shunting-yard_algorithm),
 * which is released under the
 * Creative Commons Attribution-Share-Alike License 3.0 (https://creativecommons.org/licenses/by-sa/3.0/)
 */

package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"unicode"
)

const (
	null string = "nil"
	lpa  string = "left parenthesis"
	rpa  string = "right parenthesis"
	op   string = "operation"
	i32  string = "32 bit integer"
	f32  string = "32 bit floating-point"
)

type token struct {
	sdata        string
	idata        int64
	fdata        float64
	precedence   uint
	tokenType    string
	rAssociative bool
	unary        bool
}

func (t token) toString() string {
	var s string = "("
	s += "\"" + t.sdata + "\", " + strconv.FormatInt(t.idata, 10) + ", " + fmt.Sprintf("%f", t.fdata)
	s += ", [" + strconv.FormatUint(uint64(t.precedence), 10) + " : " + string(t.tokenType)
	return s
}

type lexana struct {
	data string
}

func (l lexana) lex(s string) []token {
	var formatTokens []token
	l.data = s

	i := 0
	abort := false

	for i < len(l.data)-1 {
		c := l.data[i]
		var t token

		for c == ' ' || c == '\t' || c == '\r' {
			if l.data[i] != l.data[len(l.data)-1] {
				i++
				c = l.data[i]
			} else {
				abort = true
				break
			}
		}

		if abort {
			break
		}

		if c == '(' {
			t.tokenType = lpa
			t.sdata = "("
			t.precedence = 9
			t.rAssociative = false
		} else if c == ')' {
			t.tokenType = rpa
			t.sdata = ")"
			t.precedence = 0
			t.rAssociative = false
		} else if c == '+' {
			t.tokenType = op
			t.sdata = "+"
			t.precedence = 2
			t.rAssociative = false
		} else if c == '-' {
			ln := len(formatTokens)
			if ln < 1 {
				t.unary = true
				t.precedence = 5
				t.rAssociative = true
				t.sdata = "m"
				t.tokenType = op
			} else if formatTokens[ln-1].tokenType == op {
				t.unary = true
				t.precedence = 5
				t.rAssociative = true
				t.sdata = "m"
				t.tokenType = op
			} else if formatTokens[ln-1].tokenType == lpa {
				t.unary = true
				t.precedence = 5
				t.rAssociative = true
				t.sdata = "m"
				t.tokenType = op
			} else {
				t.tokenType = op
				t.sdata = "-"
				t.precedence = 2
				t.rAssociative = false
			}
		} else if c == '/' {
			t.tokenType = op
			t.sdata = "/"
			t.precedence = 3
			t.rAssociative = false
		} else if c == '*' || c == 'x' || c == 'X' {
			t.tokenType = op
			t.sdata = "*"
			t.precedence = 3
			t.rAssociative = false
		} else if c == '%' {
			t.tokenType = op
			t.sdata = "%"
			t.precedence = 6
			t.rAssociative = false
		} else if c == '^' {
			t.tokenType = op
			t.sdata = "^"
			t.precedence = 4
			t.rAssociative = true
		} else if unicode.IsDigit(rune(c)) {
			var floatStr string
			t.tokenType = i32

			for unicode.IsDigit(rune(c)) || c == '_' || c == '.' {
				if c == '_' {
					if l.data[i] != l.data[len(l.data)-1] {
						i++
						c = l.data[i]
					} else {
						abort = true
					}
					continue
				} else if c == '.' {
					if t.tokenType == f32 {
						fmt.Println("Redefinition of float:", t.toString())
						os.Exit(1)
					} else {
						t.tokenType = f32
						floatStr += string('.')
					}
				} else {
					floatStr += string(c)
				}

				if l.data[i] != l.data[len(l.data)-1] {
					i++
					c = l.data[i]
				} else {
					abort = true
					break
				}
			}

			if t.tokenType == f32 {
				_fdata, err := strconv.ParseFloat(floatStr, 64)
				if err == nil {
					t.fdata = _fdata
				} else {
					fmt.Println("Error: ", err)
					os.Exit(1)
				}
			} else {
				_idata, err := strconv.Atoi(floatStr)
				if err == nil {
					t.idata = int64(_idata)
				} else {
					fmt.Println("Error: ", err)
					os.Exit(1)
				}
			}

			i-- // Let for loop skip last char.
		} else if c == '.' {
			if unicode.IsDigit(rune(l.data[i+1])) {
				t.tokenType = f32
				if l.data[i] != l.data[len(l.data)-1] {
					i++
					c = l.data[i]
				} else {
					abort = true
				}
				floatStr := "."

				for unicode.IsDigit(rune(c)) || c == '_' || c == '.' {
					if c == '_' {
						if l.data[i] != l.data[len(l.data)-1] {
							i++
							c = l.data[i]
						} else {
							abort = true
						}
						continue
					} else if c == '.' {
						fmt.Println("Redefinition of float:", t.toString())
						os.Exit(1)
					} else {
						floatStr += string(c)
					}

					if l.data[i] != l.data[len(l.data)-1] {
						i++
						c = l.data[i]
					} else {
						abort = true
						break
					}
				}

				_fdata, err := strconv.ParseFloat(floatStr, 64)
				if err == nil {
					t.fdata = _fdata
				} else {
					fmt.Println("Error: ", err)
					os.Exit(1)
				}

				i-- // Let for loop skip last char.
			} else {
				fmt.Println("Unexpected: '.'")
				os.Exit(1)
			}
		} else {
			fmt.Println("Unexpected: '", c, "'")
			os.Exit(1)
		}

		formatTokens = append(formatTokens, t)

		if abort {
			break
		}

		i++
	}

	return formatTokens
}

func shuntingYard(tokens []token) []token {
	var queue []token
	var stack []token

	for _, t := range tokens {
		if t.tokenType == i32 || t.tokenType == f32 {
			queue = append(queue, t)
		} else if t.tokenType == op {
			o1 := t

			for len(stack) > 0 {
				o2 := stack[len(stack)-1]

				if (!o1.rAssociative && o1.precedence <= o2.precedence) || (o1.rAssociative && o1.precedence < o2.precedence) {
					stack = stack[:len(stack)-1]
					queue = append(queue, o2)
					continue
				}

				break
			}

			stack = append(stack, o1)
		} else if t.tokenType == lpa {
			stack = append(stack, t)
		} else if t.tokenType == rpa {
			match := false

			for len(stack) > 0 && stack[len(stack)-1].tokenType != lpa {
				queue = append(queue, stack[len(stack)-1])
				stack = stack[:len(stack)-1]
				match = true
			}

			if !match && len(stack) < 1 {
				fmt.Println("Right parenthesis error:", t.toString())
				os.Exit(1)
			}

			if len(stack) > 0 {
				stack = stack[:len(stack)-1]
			}
		}
	}

	for len(stack) > 0 {
		if stack[len(stack)-1].tokenType == lpa {
			fmt.Println("Mismatched parenthesis error")
			os.Exit(1)
		}

		queue = append(queue, stack[len(stack)-1])
		stack = stack[:len(stack)-1]
	}

	return queue
}

func compute(formatted []token) float64 {
	var stack []float64

	for len(formatted) > 0 {
		t := formatted[0]
		switch t.tokenType {
		case i32:
			stack = append(stack, float64(t.idata))

		case f32:
			stack = append(stack, t.fdata)

		case op:
			if t.unary {
				stack[len(stack)-1] = stack[len(stack)-1] * -1
			} else {
				rhs := stack[len(stack)-1]
				stack = stack[:len(stack)-1]
				lhs := stack[len(stack)-1]
				stack = stack[:len(stack)-1]

				switch t.sdata {
				case "^":
					stack = append(stack, math.Pow(lhs, rhs))

				case "*":
					stack = append(stack, lhs*rhs)

				case "/":
					stack = append(stack, lhs/rhs)

				case "+":
					stack = append(stack, lhs+rhs)

				case "-":
					stack = append(stack, lhs-rhs)
				}
			}
		}

		formatted = formatted[1:]
	}

	return stack[len(stack)-1]
}

func main() {
	var exprStr string
	lexer := new(lexana)
	scanner := bufio.NewScanner(os.Stdin)

	for {
		fmt.Print("Enter an mathematical expression ('exit' to stop): ")
		scanner.Scan()
		exprStr = scanner.Text()
		fmt.Println()

		if exprStr == "exit" {
			break
		}

		formattedTokens := lexer.lex(exprStr)

		formatted := shuntingYard(formattedTokens)

		value := compute(formatted)

		fmt.Println("That evaluates out to:\n", value)
		fmt.Println()
	}
}
